src/dwn/models/record.js

```diff
@@ -0,0 +1,244 @@
+      // `readableStream` will be set if Record was instantiated from a RecordsRead reply.
+      // If neither of the above are true, then the record must be fetched from the DWN.
+      this.#readableStream = this.#dwn.records.read(this.#target, { author: this.#author, message: { recordId: this.#recordId } })
+        .then((response) => response.record )
+        .then((record) => { return record.data; });
+    }
+
+    if (typeof this.#encodedData === 'string') {
+      // If `encodedData` is set, then it is expected that:
+      // `dataSize` <= DwnConstant.maxDataSizeAllowedToBeEncoded (10KB as of April 2023)
+      // type is Uint8Array bytes if the Record object was instantiated from a RecordsWrite response
+      // type is Base64 URL encoded string if the Record object was instantiated from a RecordsQuery response
+      // If it is a string, we need to Base64 URL decode to bytes
+      this.#encodedData = Encoder.base64UrlToBytes(this.#encodedData);
+    }
+
+    const self = this; // Capture the context of the `Record` instance.
+    const dataObj = {
+      async json() {
+        if (self.#encodedData) return this.text().then(JSON.parse);
+        if (self.#readableStream) return this.text().then(JSON.parse);
+        return null;
+      },
+      async text() {
+        if (self.#encodedData) return Encoder.bytesToString(self.#encodedData);
+        if (self.#readableStream) return self.#readableStream.then(DataStream.toBytes).then(Encoder.bytesToString);
+        return null;
+      },
+      async stream() {
+        if (self.#encodedData) return DataStream.fromBytes(self.#encodedData);
+        if (self.#readableStream) return self.#readableStream;
+        return null;
+      },
+      then (...callbacks) {
+        return this.stream().then(...callbacks);
+      },
+      catch(callback) {
+        return dataObj.then().catch(callback);
+      },
+    };
+    return dataObj;
+  }
+
+  /**
+   * Record mutation methods.
+   */
+
+  async delete() {
+    if (this.isFrozen) throw new Error(`Error: Record with ID '${this.id}' was previously deleted.`);
+
+    // Attempt to delete the record from the DWN.
+    const response = await this.#dwn.records.delete(this.#target, { author: this.#author, message: { recordId: this.#recordId } });
+    
+    if (response.status.code === 202) {
+      // If the record was successfully deleted, freeze the instance to prevent further modifications.
+      this.#freezeRecord();
+    }
+
+    return response;
+  }
+
+  async update(options = { }) {
+    if (this.isFrozen) throw new Error(`Error: Record with ID '${this.id}' was previously deleted.`);
+
+    // Begin assembling update message.
+    let updateMessage = { ...this.#descriptor, ...options };
+
+    // If `data` is being updated then `dataCid` and `dataSize` must be undefined and the `data` property is passed as
+    // a top-level property to `web5.dwn.records.write()`.
+    let data;
+    if (options?.data !== undefined) {
```

Comment:
are we allowing for a scenario where `options` doesnt exist even though we've given it a default value?

AFAIK the only way for that to be possible is if the caller explicitly provides `null` or a non-object value (in which case we probably should be checking for that, no?)

Reply:
Addressed in 9870e2fc5d088c1ad216eeecefd0d81b23d0e8f8